ELEX-3031: Potential Voterflow Solvers

dmnapolitano opened the pull request:
## Description

This PR provides ~~three~~ one potential transition solver and a base class 🎉 

Work on bootstrapping the matrix solver is still in-progress and can be found in [this branch](https://github.com/washingtonpost/elex-solver/tree/ELEX-3830-bootstrap-matrix-solver).

The EI solver is now the EI model and is part of [elex-voterflow-model](https://github.com/WPMedia/elex-voterflow-model), where it is still a subclass of `TransitionSolver`.

## Jira Ticket

ELEX-3031

## Test Steps

`tox`!

dmnapolitano commented:
Hi @lennybronner 👋🏻 

Alright, I think this is officially ready for review!  Looking forward to addressing all your questions/comments/concerns 😄 🎉 

In particular, please take a look at how I implemented weighting in the matrix solver.  I'm curious if I got something wrong there and that's why we're having problems with weighting + stratifying 🤔 

Thanks!!

lennybronner commented on src/elexsolver/TransitionMatrixSolver.py:
```diff
@@ -1,28 +1,107 @@
+        self._check_any_element_nan_or_inf(Y)
+
+        # matrices should be (units x things), where the number of units is > the number of things
+        if X.shape[1] > X.shape[0]:
```
I'm not sure we should force this to be true. Not really true for our use-case but the optimization should still work if p > n

dmnapolitano replied:
It will, it will just transpose `X` (and then `Y` right below this), although I realize that might not be what everyone wants to happen.  I added in logic like this because I was getting really confused about the expected dimensions of the data and getting a lot of errors, so I added in code like this to standardize everything one way (units x things (candidates)).

The other issue is, if p > n, your result is going to be horrible if the solver even succeeds 😬  However, this might be something that's specific to voterflow and possibly other, but not all, uses of this solver, so I can move `_check_dimensions()` out of here and into `elex-voterflow-model`. 👍🏻 

But I guess even this check and transpose here on line 79 might also be specific to the use of the solver and not something inherently important to the solver itself 🤔

Ok, I'll try moving this to `elex-voterflow-model` as well and see what happens here 😬 😄

lennybronner commented on src/elexsolver/TransitionMatrixSolver.py:
```diff
@@ -1,28 +1,107 @@
-        return A @ self.transition_matrix
+        percentages = self.__solve(X, Y, weights)
+        self._transitions = np.diag(X_expected_totals) @ percentages
+        return percentages
```
doesn't `fit_predict` usually return the labels? So in this case `self._transitions` rather than `percentages`? Is there any reason you didn't implement `fit` and `predict` and then have `fit_predict` call the two functions?

dmnapolitano replied:
Great questions that I will answer in reverse order:

1. I went with `fit_predict()` instead of `fit()` followed by `predict()` because in `scikit-learn` convention, one can `fit()` on one data set and then at some later point `predict()` on an entirely different data set (with the same features).  Here, the `predict()` has to happen on the same data set as the `fit()`, so I figured I'd keep with the `scikit-learn` convention for solvers where this is the case (e.g. clustering).  Also the `predict()` here would have to take nothing as an argument, which would break with `scikit-learn`'s `predict()` should we ever want to use anything from there e.g. grid search down the line. 🤷🏻‍♀️ 

2. This might be because my naming convention here is absolutely terrible so _please_ suggest new names 🙈  The transitions are the "flow-y bits" that get used in sankey diagrams.  If you sum the entire `self._transitions` matrix it should sum to 1 (100%).  However, the `percentages` are the beta terms that come out of the solver such that each row (representing a "from" candidate) sums to 1 (100%).  And only the `percentages` are being returned to again follow `scikit-learn` conventions where `fit_predict()` returns one thing, the predictions.

lennybronner replied:
Thanks for the detailed answers!
1. This solver is more general than the voterflow problem that we are using it for and there may be situations in which we (or someone else) will want to predict on data that was not used for fitting (for example if this were part of an election model, where we fit on a subset of the data but apply it to the full dataset).

2. The beta terms are not the predictions. The solution to our problem might be the coefficients of the optimization problem, but the predictions are still `y_hat = \hat{\beta} @ X`

dmnapolitano replied:
Oh, I see!  I'm so sorry, it never occurred to me that there might be multiple use-cases for this! 😓 

Ok, in that case I will:

* change this from `fit_predict()` to `fit()` and `predict()` where `fit()` returns nothing and `predict()` returns `Y_hat`;
* create another class member for the "percentages", probably just called `_betas` since that's more obvious lol;

Two questions though:

1. Does it make sense to keep calling this `Transition(Matrix)Solver`?
2. Does it make sense to do the math here for the `_transitions` or is that something exclusive to voterflow?  I do have spots in `elex-voterflow-model` where I do that math again (`_handle_non_voters()`) so if this is 💯 a voterflow thing, I can remove it from here and just keep it there.

🤔

lennybronner replied:
1. I'm not sure I follow why it wouldn't still be `TransitionMatrixSolver`?
2. Isn't computing transition what we'll be doing in the predict function?

dmnapolitano replied:
1. Yeah, thinking this through more, maybe you're right and the name works.  I was just thinking that, if `predict()` returns `Y_hat` instead of `betas`, which are what I've been calling the "percentages" and I think you're calling the "transitions", what is the goal of this solver?  Is it `Y_hat` or the `betas`?  But really, it's both, like with any other regression, really 🤔 
2. No, `predict()` will compute and return `Y_hat`.  This is awful terminology on my part and I'm so sorry 😬  You can't use the `betas` in the Sankey diagrams because they'll create a diagram that sums to more than 100% and displays the wrong per-candidate percentage totals on either side.  So we have to do math to create "transitions" or the flow-y pieces, as in: https://github.com/washingtonpost/elex-solver/blob/ELEX-3031-voterflow-models/src/elexsolver/TransitionMatrixSolver.py#L97 and https://github.com/WPMedia/elex-voterflow-model/blob/ELEX-3031-voterflow-models/src/elexvoterflow/voterflow.py#L198.  So what I'm wondering is, is this math to convert `betas` into the flow-y pieces something unique to the voterflow use-case?  If so, I'll remove it from the solver 🤔

lennybronner commented on src/elexsolver/TransitionSolver.py:
```diff
@@ -0,0 +1,111 @@
+        """
+        raise NotImplementedError
+
+    def get_prediction_interval(self, pi: float):
```
do we need this function?

dmnapolitano replied:
Good catch!  I'll remove it.  I didn't get that far lol 😅

lennybronner commented on src/elexsolver/TransitionSolver.py:
```diff
@@ -0,0 +1,111 @@
+LOG = logging.getLogger(__name__)
+
+
+class TransitionSolver(ABC):
```
is there any way/ does it make sense for this to inherit from LinearSolver? Then we'd get a lot of the checking functions for free?

dmnapolitano replied:
Interesting idea!  I'll try it!  Thanks! 🤔 🤞🏻

dmnapolitano replied:
Ok, I just pushed changes where `TransitionSolver` inherits from `LinearSolver` 🎉  Curious to know your thoughts 🤔

lennybronner reviewed (commented):
This looks great! I left some small notes/questions.

lennybronner commented on tests/test_transition_matrix_solver.py:
```diff
@@ -0,0 +1,230 @@
+
+    weights = np.array([500, 250, 125, 62.5, 31.25, 15.625])
+
+    expected = np.array([[0.737329, 0.262671], [0.230589, 0.769411]])
```
I'm curious where this expected is from?

dmnapolitano replied:
It's the result of running the code on the next few lines below this.  Do you get a different result if you run this locally? 😬

lennybronner replied:
no, I was just curious. This isn't super ideal (for quantile regression the expected solution was taken from an implementation in R) but I guess we have no other option here.

dmnapolitano commented:
Hi @lennybronner I removed the requirement of everything having to be integers here and updated the unit tests 🎉  Let me know what you think and I'll make sure `elex-voterflow-model` is working with these changes.  Thanks! 😄
